mqtt/policy/src/core/mod.rs

```diff
@@ -843,4 +833,98 @@ pub(crate) mod tests {
             policy.starts_with(input)
         }
     }
+
+    #[cfg(feature = "proptest")]
+    mod proptests {
+        use crate::{Decision, Effect, PolicyBuilder, PolicyDefinition, Request, Statement};
+        use proptest::{collection::vec, prelude::*};
+
+        proptest! {
+            /// The goal of this test is to verify the following scenarios:
+            /// - PolicyBuilder does not crash.
+            /// - All combinations of identity/operation/resource in a statement in the definition
+            ///   should produce expected result.
+            ///   Since some statements can be overridden by the previous ones,
+            ///   we can only safely verify the very first statement.
+            #[test]
+            fn policy_engine_proptest(definition in arb_policy_definition()){
+                use itertools::Itertools;
+
+                // take very first statement, which should have top priority.
+                let statement = &definition.statements()[0];
```

Comment:
I guess it is possible to receive an empty definition from the customer. Do we want to cover that here?

Reply:
It is possible, and in that case every request will return default decision. We don't need to cover that in proptests though, I believe we have a unit test for that.